File: internal/proxy/views/custom/panels/scripts.js
Ext.define("PBS.D2DManagement.ScriptPanel", {
  extend: "Ext.grid.Panel",
  alias: "widget.pbsDiskScriptPanel",

  stateful: true,
  stateId: "grid-disk-backup-scripts-v1",

  controller: {
    xclass: "Ext.app.ViewController",

    onAdd: function() {
      let me = this;
      Ext.create("PBS.D2DManagement.ScriptEditWindow", {
        listeners: {
          destroy: function() {
            me.reload();
          },
        },
      }).show();
    },

    onEdit: function() {
      let me = this;
      let view = me.getView();
      let selection = view.getSelection();
      if (!selection || selection.length < 1) {
        return;
      }
      Ext.create("PBS.D2DManagement.ScriptEditWindow", {
        contentid: selection[0].data.path,
        autoLoad: true,
        listeners: {
          destroy: () => me.reload(),
        },
      }).show();
    },

    removeScripts: function() {
      const me = this;
      const view = me.getView();
      const recs = view.getSelection();
      if (!recs.length) return;

      Ext.Msg.confirm(
        gettext("Confirm"),
        gettext("Remove selected entries?"),
        (btn) => {
          if (btn !== "yes") return;
          recs.forEach((rec) => {
            PBS.PlusUtils.API2Request({
              url:
                "/api2/extjs/config/d2d-script/" +
                encodeURIComponent(encodePathValue(rec.getId())),
              method: "DELETE",
              waitMsgTarget: view,
              failure: (resp) =>
                Ext.Msg.alert(gettext("Error"), resp.htmlStatus),
              success: () => me.reload(),
            });
          });
        }
      );
    },

    reload: function() {
      this.getView().getStore().rstore.load();
    },

    stopStore: function() {
      this.getView().getStore().rstore.stopUpdate();
    },

    startStore: function() {
      this.getView().getStore().rstore.startUpdate();
    },

    init: function(view) {
      Proxmox.Utils.monStoreErrors(view, view.getStore().rstore);
    },
  },

  listeners: {
    beforedestroy: "stopStore",
    deactivate: "stopStore",
    activate: "startStore",
    itemdblclick: "onEdit",
  },

  store: {
    type: "diff",
    rstore: {
      type: "update",
      storeid: "proxmox-disk-scripts",
      model: "pbs-model-scripts",
      proxy: {
        type: "pbsplus",
        url: pbsPlusBaseUrl + "/api2/json/d2d/script",
      },
    },
    sorters: "path",
  },

  features: [
  ],

  tbar: [
    {
      text: gettext("Add"),
      xtype: "proxmoxButton",
      handler: "onAdd",
      selModel: false,
    },
    "-",
    {
      text: gettext("Edit"),
      xtype: "proxmoxButton",
      handler: "onEdit",
      disabled: true,
    },
    {
      xtype: "proxmoxButton",
      text: gettext("Remove"),
      handler: "removeScripts",
      enableFn: function() {
        let recs = this.up("grid").getSelection();
        return recs.length > 0;
      },
      disabled: true,
    },
  ],
  columns: [
    {
      text: gettext("Path"),
      dataIndex: "path",
      flex: 2,
    },
    {
      text: gettext("Description"),
      dataIndex: "description",
      flex: 1,
    },
    {
      text: gettext("Job Count"),
      dataIndex: "job_count",
      flex: 1,
    },
    {
      text: gettext("Target Count"),
      dataIndex: "target_count",
      flex: 1,
    },
  ],
});

File: internal/server/web/ui/management/scripts.go
package management

import (
	"github.com/pbs-plus/pbs-plus/internal/server/web/js"
)

const scriptAPIPath = "/api2/json/d2d/script"

var scriptModel = js.Model{
	Name:       "pbs-model-scripts",
	Fields:     js.Fields("path", "description", "job_count", "target_count"),
	IDProperty: "path",
}

var scriptSelector = js.Selector{
	Name:         "PBS.form.D2DScriptSelector",
	XType:        "pbsD2DScriptSelector",
	DisplayField: "path",
	ValueField:   "path",
	APIPath:      scriptAPIPath,
	Sorters:      "path",
	AllowBlank:   new(true),
	AutoSelect:   new(false),
	ListWidth:    450,
	ListColumns: []js.Column{
		{Text: "Path", DataIndex: "path", Flex: 3, Sortable: new(true), Renderer: "Ext.String.htmlEncode"},
		{Text: "Description", DataIndex: "description", Flex: 3, Sortable: new(true), Renderer: "Ext.String.htmlEncode"},
	},
}

var scriptPanel = js.Panel{
	Name:    "PBS.D2DManagement.ScriptPanel",
	XType:   "pbsDiskScriptPanel",
	StateID: "grid-disk-backup-scripts-v1",
	Store: js.Store{
		StoreID: "proxmox-disk-scripts",
		Model:   "pbs-model-scripts",
		APIPath: scriptAPIPath,
		Sorters: "path",
	},
	Listeners: js.Listeners{ItemDblClick: "onEdit"},
	Controller: js.Controller{Methods: map[string]js.Raw{
		"onAdd":  js.OpenEditWindow("PBS.D2DManagement.ScriptEditWindow", ""),
		"onEdit": js.OpenEditWindow("PBS.D2DManagement.ScriptEditWindow", "path"),
		"removeScripts": js.ConfirmRemove(
			"/api2/extjs/config/d2d-script/",
			"encodePathValue(rec.getId())",
			"Remove selected entries?",
		),
	}},
	Tbar: []js.Tool{
		{Text: "Add", Handler: "onAdd", SelModel: new(false)},
		js.Sep(),
		{Text: "Edit", Handler: "onEdit", Disabled: true},
		{Text: "Remove", Handler: "removeScripts", Disabled: true, EnableFn: js.EnableOnSelection},
	},
	Columns: []js.Column{
		{Text: "Path", DataIndex: "path", Flex: 2},
		{Text: "Description", DataIndex: "description", Flex: 1},
		{Text: "Job Count", DataIndex: "job_count", Flex: 1},
		{Text: "Target Count", DataIndex: "target_count", Flex: 1},
	},
}

var scriptEditWindow = js.EditWindow{
	Name:      "PBS.D2DManagement.ScriptEditWindow",
	XType:     "pbsScriptEditWindow",
	Subject:   "Script",
	Width:     "80%",
	Resizable: true,
	IsCreate:  true,
	IsAdd:     true,
	CBindData: js.PathKeyedURL("/api2/extjs/config/d2d-script"),
	Items: js.Items(
		js.Field{
			XType:      js.XDisplayEditField,
			Label:      "Description",
			Name:       "description",
			Renderer:   "Ext.htmlEncode",
			AllowBlank: new(true),
			Editable:   new(true),
		},
		js.Field{
			XType:  js.XFieldContainer,
			Label:  "Script Content",
			Layout: "anchor",
			Items:  js.Items(js.CodeMirrorField("script", "shell", 400)),
		},
	),
	Methods: map[string]js.Raw{
		"getValues": js.Func("", `
			let values = this.callParent();
			let editor = this.down("#scriptEditor");
			if (editor && editor.codeMirror) {
				values.script = editor.codeMirror.getValue();
			}
			return values;
		`),
		"setValues": js.Func("values", `
			this.callParent([values]);
			let editor = this.down("#scriptEditor");
			if (editor && editor.codeMirror && values.script) {
				editor.codeMirror.setValue(values.script);
			}
		`),
	},
}
